tests/beta/interfaces/test_tape_tf.py

```diff
@@ -0,0 +1,741 @@
+            res = qtape.execute(dev)
+
+        assert isinstance(res, tf.Tensor)
+        assert res.shape == (2,)
+
+        expected = [tf.cos(a), -tf.cos(a) * tf.sin(b)]
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+
+        res = tape.jacobian(res, [a, b])
+        expected = [[-tf.sin(a), tf.sin(a) * tf.sin(b)], [0, -tf.cos(a) * tf.cos(b)]]
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+
+        spy.assert_called()
+
+    def test_jacobian_dtype(self, tol):
+        """Test calculating the jacobian with a different datatype"""
+        a = tf.Variable(0.1, dtype=tf.float32)
+        b = tf.Variable(0.2, dtype=tf.float32)
+
+        dev = qml.device("default.qubit", wires=2)
+
+        with tf.GradientTape() as tape:
+            with TFInterface.apply(QuantumTape(), dtype=tf.float32) as qtape:
+                qml.RY(a, wires=0)
+                qml.RX(b, wires=1)
+                qml.CNOT(wires=[0, 1])
+                expval(qml.PauliZ(0))
+                expval(qml.PauliY(1))
+
+            assert qtape.trainable_params == {0, 1}
+            res = qtape.execute(dev)
+
+        assert isinstance(res, tf.Tensor)
+        assert res.shape == (2,)
+        assert res.dtype is tf.float32
+
+        res = tape.jacobian(res, [a, b])
+        assert [r.dtype is tf.float32 for r in res]
+
+    def test_jacobian_options(self, mocker, tol):
+        """Test setting jacobian options"""
+        spy = mocker.spy(QuantumTape, "numeric_pd")
+
+        a = tf.Variable([0.1, 0.2])
+
+        dev = qml.device("default.qubit", wires=1)
+
+        with tf.GradientTape() as tape:
+            with TFInterface.apply(QuantumTape()) as qtape:
+                qml.RY(a[0], wires=0)
+                qml.RX(a[1], wires=0)
+                expval(qml.PauliZ(0))
+
+            res = qtape.execute(dev)
+
+        qtape.jacobian_options = {"h": 1e-8, "order": 2}
+        tape.jacobian(res, a)
+
+        for args in spy.call_args_list:
+            assert args[1]["order"] == 2
+            assert args[1]["h"] == 1e-8
+
+    def test_reusing_quantum_tape(self, tol):
+        """Test re-using a quantum tape by passing new parameters"""
+        a = tf.Variable(0.1, dtype=tf.float64)
+        b = tf.Variable(0.2, dtype=tf.float64)
+
+        dev = qml.device("default.qubit", wires=2)
+
+        with tf.GradientTape() as tape:
+
+            with TFInterface.apply(QuantumTape()) as qtape:
+                qml.RY(a, wires=0)
+                qml.RX(b, wires=1)
+                qml.CNOT(wires=[0, 1])
+                expval(qml.PauliZ(0))
+                expval(qml.PauliY(1))
+
+            assert qtape.trainable_params == {0, 1}
+
+            res = qtape.execute(dev)
+
+        jac = tape.jacobian(res, [a, b])
+
+        a = tf.Variable(0.54, dtype=tf.float64)
+        b = tf.Variable(0.8, dtype=tf.float64)
+
+        with tf.GradientTape() as tape:
+            res2 = qtape.execute(dev, params=[2 * a, b])

```

Comment:
Very minor thing: would the order of names `jac` and `res2` be reversed here? :smile:

Reply:
fixed!